pulsar/consumer_zero_queue.go

```diff
@@ -0,0 +1,328 @@
+	closeCh   chan struct{}
+	errorCh   chan error
+
+	log     log.Logger
+	metrics *internal.LeveledMetrics
+}
+
+func newZeroConsumer(client *client, options ConsumerOptions, topic string,
+	messageCh chan ConsumerMessage, dlq *dlqRouter,
+	rlq *retryRouter, disableForceTopicCreation bool) (*zeroQueueConsumer, error) {
+	zc := &zeroQueueConsumer{
+		topic:                     topic,
+		client:                    client,
+		options:                   options,
+		disableForceTopicCreation: disableForceTopicCreation,
+		messageCh:                 messageCh,
+		closeCh:                   make(chan struct{}),
+		errorCh:                   make(chan error),
+		dlq:                       dlq,
+		rlq:                       rlq,
+		log:                       client.log.SubLogger(log.Fields{"topic": topic}),
+		consumerName:              options.Name,
+		metrics:                   client.metrics.GetLeveledMetrics(topic),
+	}
+
+	var nackRedeliveryDelay time.Duration
+	if zc.options.NackRedeliveryDelay == 0 {
+		nackRedeliveryDelay = defaultNackRedeliveryDelay
+	} else {
+		nackRedeliveryDelay = zc.options.NackRedeliveryDelay
+	}
+	opts := &partitionConsumerOpts{
+		topic:                       zc.topic,
+		consumerName:                zc.consumerName,
+		subscription:                zc.options.SubscriptionName,
+		subscriptionType:            zc.options.Type,
+		subscriptionInitPos:         zc.options.SubscriptionInitialPosition,
+		partitionIdx:                0,
+		receiverQueueSize:           zc.options.ReceiverQueueSize,
+		nackRedeliveryDelay:         nackRedeliveryDelay,
+		nackBackoffPolicy:           zc.options.NackBackoffPolicy,
+		metadata:                    zc.options.Properties,
+		subProperties:               zc.options.SubscriptionProperties,
+		replicateSubscriptionState:  zc.options.ReplicateSubscriptionState,
+		startMessageID:              zc.options.startMessageID,
+		startMessageIDInclusive:     zc.options.StartMessageIDInclusive,
+		subscriptionMode:            zc.options.SubscriptionMode,
+		readCompacted:               zc.options.ReadCompacted,
+		interceptors:                zc.options.Interceptors,
+		maxReconnectToBroker:        zc.options.MaxReconnectToBroker,
+		backoffPolicy:               zc.options.BackoffPolicy,
+		keySharedPolicy:             zc.options.KeySharedPolicy,
+		schema:                      zc.options.Schema,
+		decryption:                  zc.options.Decryption,
+		ackWithResponse:             zc.options.AckWithResponse,
+		maxPendingChunkedMessage:    zc.options.MaxPendingChunkedMessage,
+		expireTimeOfIncompleteChunk: zc.options.ExpireTimeOfIncompleteChunk,
+		autoAckIncompleteChunk:      zc.options.AutoAckIncompleteChunk,
+		consumerEventListener:       zc.options.EventListener,
+		enableBatchIndexAck:         zc.options.EnableBatchIndexAcknowledgment,
+		ackGroupingOptions:          zc.options.AckGroupingOptions,
+		autoReceiverQueueSize:       zc.options.EnableAutoScaledReceiverQueueSize,
+	}
```

Comment:
Could we abstract this logic so that we can reuse the code? https://github.com/apache/pulsar-client-go/blob/fc15995c1a180a151aca58c478bc40843120783e/pulsar/consumer_impl.go#L390-L420

Reply:
Thanks, done.